code/ecdsap256sha2/Hacl.Impl.SolinasReduction.fst

```diff
@@ -0,0 +1,593 @@
+module Hacl.Impl.SolinasReduction
+
+open FStar.HyperStack.All
+open FStar.HyperStack
+module ST = FStar.HyperStack.ST
+
+open Lib.IntTypes 
+open FStar.Math.Lemmas
+open FStar.Math.Lib
+open Lib.Buffer
+
+open Hacl.Spec.P256.SolinasReduction
+open Hacl.Impl.LowLevel
+open Hacl.Impl.P256.LowLevel
+
+open Hacl.Spec.P256.Definitions
+open FStar.Mul
+
+
+module Seq = Lib.Sequence
+
+inline_for_extraction noextract
+val get_high_part: a: uint64 -> Tot (r: uint32{uint_v r == uint_v a / (pow2 32)})
+
+let get_high_part a = 
+  to_u32(shift_right a (size 32))
+
+inline_for_extraction noextract
+val get_low_part: a: uint64 -> Tot (r: uint32{uint_v r == uint_v a % (pow2 32)}) 
+
+let get_low_part a = to_u32 a
+
+val lemma_xor_zero: low: uint64{uint_v (get_high_part low) ==  0} -> high: uint64{uint_v (get_low_part high) == 0} ->  Lemma (uint_v (logxor low high) = uint_v high * pow2 32 + uint_v low)
```

Comment:
This assumed lemma doesn't hold in general and one could prove `False` from it.

The statement should read:
```FStar
val lemma_xor_zero: 
    low:  uint64{uint_v (get_high_part low) == 0} 
  -> high: uint64{uint_v (get_low_part high) == 0} 
  -> Lemma (uint_v (logxor low high) = uint_v high + uint_v low)
```

Reply:
Fixed